src/lightning/fabric/strategies/deepspeed.py

```diff
@@ -831,3 +833,14 @@ def _validate_state_keys(state: Dict[str, Any]) -> None:
             " values being overwritten by DeepSpeed. Consider changing the name of these keys to something else: "
             + ", ".join(colliding_keys)
         )
+
+
+def _validate_device_index_selection(parallel_devices: List[torch.device]) -> None:
+    selected_device_indices = [device.index for device in parallel_devices]
+    expected_device_indices = list(range(len(parallel_devices)))
```

Comment:
If order doesnt matter then this should use sets

Reply:
The order does matter indeed :)  If user passed [0, 2, 1], the ranks for the processes would still be [0, 1, 2] and deepspeed would access the GPUs as [0, 1, 2], contradicting what the user asked for.

Reply:
Shouldnt you use sets then in case parallel devices is not sorted?

Reply:
I already have this covered in the test case with [1, 0]. If we used set comparison here, then [0, 1] = [1, 0] and then it wouldn't raise an error.